### DISCUSSION TITLE
How to manually unregister a field when shouldUnregister is set to false

### DISCUSSION THREAD

**Discussion by AleKrabbe** [react-hook-form/react-hook-form#2627] [Q&A]
Hello,

I have a form with `shouldUnregister` set to `false`, because I render fields conditionally, and sometimes a field that was previously rendered and therefore registered, becomes hidden (`unmounted`) by a user action, nonetheless I still need that hidden field value on the form data; so `shouldUnregister` set to `false` solves this problem. (as @bluebill1049 answered on my last post)

Now I'm having trouble when another user action causes a field that was previously registered to be permanently deleted, so I dont need nor want the deleted field value on the submitted form data. At first I though that using the `unregister` would solve my problem here; however, since I have `shouldUnregister` set to `false`, that is not the case. The deleted field remains registered and still gets validated by my `yupResolver`.

How can I got about having `shouldUnregister` set to false and still be able to unregister fields manually if needed?

Thank you!

**Comment by bluebill1049** [comment]
you can't unregister at `shouldUnregister:false` just like the controlled local state, but you can `setValue` to `''`

**Reply by AleKrabbe** [reply]
Thanks for the reply!

The problem here is that if I `setValue` to `''` the field will still go through validation, and the field cannot be empty, so I get a validation error on a field that should have been deleted (`unregistered`) in the first place.

Do you have any suggestions / workaround? Do you think adding a feature where a user could manually `unregister` a field, eventhough `shouldUnregister` is set to `false`, would break anything?

### EDIT

I say that because from the docs I understand that fields are automatically unregistered when unmounted, and `shouldUnregister` prevents that, so I don't see a problem on allowing a manual override of that feature by letting the user manually using  `unregister`.

**Reply by mousaa** [reply]
I am running into the same use case too - did you end up finding a workaround?

**Reply by codeashian** [reply]
Also having this issue. Is there a workaround for this? I need to use `shouldUnregister: false` but I also need to be able to manually unregister some fields.

**Reply by patthsu** [reply]
Guessing there's no solution or workaround? 

Edit:

Sort of figured out a workaround.. I recreate the form value in a new object removing the fields that I do not want, and I use a reset to feed in the new form value. It's not optimal but seems to work..

**Reply by spirit-team** [reply]
The workaround is probably to have conditional validation on the field you want to unregister. Either through a state value or by using a useWatch

```
const watchTest = useWatch({
    name: `test`,
    defaultValue: getValues(`test`),
    control,
  });

<input
    name: `test2`,
    defaultValue: getValues(`test2`),
    ref={register}
    validate={watchTest === "Dont validate test2" ? null : {required: "test2 is required"}}
/>
```

**Reply by wdfinch** [reply]
V7 will change this so form values will be kept in form state by default and can be unregistered.

**Reply by bluebill1049** [reply]
yea, we are trying to address this issue.

**Reply by FarisMarouane** [reply]
@AleKrabbe  In response to your edit: "By default, an input value will be retained when input is removed. " (v7)
